File: jps-ui/src/aabb.cpp
// Copyright © 2024 Forschungszentrum Jülich GmbH
// SPDX-License-Identifier: LGPL-3.0-or-later
#include "aabb.hpp"

float AABB::Aspect() const
{
    const auto diff = max - min;
    return diff.x / diff.y;
}

float AABB::Width() const
{
    return max.x - min.x;
}

float AABB::Height() const
{
    return max.y - min.y;
}
glm::vec2 AABB::Center() const
{
    return min + (max - min) / 2.0f;
}

bool AABB::Contains(const glm::vec2& p) const
{
    return p.x >= min.x && p.x <= max.x && p.y >= min.y && p.y <= max.y;
}

AABB CreateFromPoints(const std::vector<glm::vec2>& points)
{
    float xMin = std::numeric_limits<float>::max();
    float xMax = std::numeric_limits<float>::lowest();
    float yMin = std::numeric_limits<float>::max();
    float yMax = std::numeric_limits<float>::lowest();

    for(const auto& p : points) {
        xMin = std::min(xMin, p.x);
        xMax = std::max(xMax, p.x);
        yMin = std::min(yMin, p.y);
        yMax = std::max(yMax, p.y);
    }

    return AABB{{xMin, yMin}, {xMax, yMax}};
}

File: python_modules/jupedsim/jupedsim/internal/aabb.py
# Copyright © 2012-2024 Forschungszentrum Jülich GmbH
# SPDX-License-Identifier: LGPL-3.0-or-later
class AABB:
    def __init__(
        self, *, xmin: float, xmax: float, ymin: float, ymax: float
    ) -> None:
        if xmax < xmin or ymax < ymin:
            raise Exception(
                f"Invalid arguments to create AABB: max values have to be larger than min values."
            )
        self.xmin = xmin
        self.xmax = xmax
        self.ymin = ymin
        self.ymax = ymax

    @property
    def width(self) -> float:
        return abs(self.xmax - self.xmin)

    @property
    def height(self) -> float:
        return abs(self.ymax - self.ymin)

    @property
    def center(self) -> tuple[float, float]:
        return (self.xmin + self.width / 2, self.ymin + self.height / 2)

    @staticmethod
    def combine(aabb, *other) -> "AABB":
        xmin = aabb.xmin
        ymin = aabb.ymin
        xmax = aabb.xmax
        ymax = aabb.ymax

        for aabb in other:
            xmin = min(xmin, aabb.xmin)
            ymin = min(ymin, aabb.ymin)
            xmax = max(xmax, aabb.xmax)
            ymax = max(ymax, aabb.ymax)

        return AABB(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax)
